src/main/java/dev/sigstore/KeylessSigner.java

```diff
@@ -0,0 +1,185 @@
+                .build(),
+            RekorVerifier.newRekorVerifier(rekorPublicKey))
+        .oidcClient(WebOidcClient.builder().setIssuer(WebOidcClient.STAGING_DEX_ISSUER).build())
+        .signer(Signers.newEcdsaSigner());
+    return builder;
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public static class Builder {
+    private FulcioClient fulcioClient;
+    private FulcioVerifier fulcioVerifier;
+    private RekorClient rekorClient;
+    private RekorVerifier rekorVerifier;
+    private OidcClient oidcClient;
+    private Signer signer;
+
+    public Builder fulcioClient(FulcioClient fulcioClient, FulcioVerifier fulcioVerifier) {
+      this.fulcioClient = fulcioClient;
+      this.fulcioVerifier = fulcioVerifier;
+      return this;
+    }
+
+    public Builder rekorClient(RekorClient rekorClient, RekorVerifier rekorVerifier) {
+      this.rekorClient = rekorClient;
+      this.rekorVerifier = rekorVerifier;
+      return this;
+    }
+
+    public Builder oidcClient(OidcClient oidcClient) {
+      this.oidcClient = oidcClient;
+      return this;
+    }
+
+    public Builder signer(Signer signer) {
+      this.signer = signer;
+      return this;
+    }
+
+    public KeylessSigner build() {
+      Preconditions.checkNotNull(fulcioClient);
+      Preconditions.checkNotNull(fulcioVerifier);
+      Preconditions.checkNotNull(rekorClient);
+      Preconditions.checkNotNull(rekorVerifier);
+      Preconditions.checkNotNull(oidcClient);
+      Preconditions.checkNotNull(signer);
+      return new KeylessSigner(
+          fulcioClient, fulcioVerifier, rekorClient, rekorVerifier, oidcClient, signer);
+    }
+  }
+
+  public void sign(Path artifact)
+      throws OidcException, NoSuchAlgorithmException, SignatureException, InvalidKeyException,
+          UnsupportedAlgorithmException, SerializationException, CertificateException, IOException,
+          FulcioVerificationException, RekorVerificationException {
+    var tokenInfo = oidcClient.getIDToken();
+    var signingCert =
+        fulcioClient.SigningCert(
+            CertificateRequest.newCertificateRequest(
+                signer.getPublicKey(),
+                tokenInfo.getIdToken(),
+                signer.sign(
+                    tokenInfo.getSubjectAlternativeName().getBytes(StandardCharsets.UTF_8))));
+    fulcioVerifier.verifyCertChain(signingCert);
+    // TODO: this signing workflow mandates SCTs, but fulcio itself doesn't, figure out a way to
+    // allow that to be known
+    fulcioVerifier.verifySct(signingCert);
+
+    // TODO: this reads the whole artifact which is probably expensive for large files, the signer
+    // interface should probably just sign non-digest artifacts instead of always applying sha256
+    // before signing. Then we can just apply the input stream to the sha256 calculator
+    // and send that digest to the signer
+    var artifactBytes = Files.readAllBytes(artifact);
```

Comment:
this should also be a stream shouldn't it?

Reply:
What do you mean? IF you mean inputstream then yes, I need to actually do the right thing here. That's the whole TODO.